### PR TITLE
server: add gRPC rate limit

### LINKED ISSUE #5739: Provide PD self-protection for gRPC requests
## Enhancement Task
We have encountered some cases that create unexcepted numbers of gRPC requests, which greatly affect the PD performance. We need to prevent this kind of thing from happening.

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: ref #6556. close #5739.

### What is changed and how does it work?

This PR adds a gRPC rate limit for some interfaces to prevent wasting lots of resources.

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: errors.toml [hunk 1]
```diff
@@ -701,6 +701,11 @@ error = '''
 leader is nil
 '''
 
+["PD:server:ErrReachRateLimit"]
```
Comment (nolouch) [inline errors.toml:704]:
```suggestion
["PD:server:ErrRateLimitExceeded"]
```

File: server/grpc_service.go [hunk 2]
```diff
@@ -756,6 +768,17 @@ func (s *GrpcServer) IsSnapshotRecovering(ctx context.Context, request *pdpb.IsS
...
 // GetStore implements gRPC PDServer.
 func (s *GrpcServer) GetStore(ctx context.Context, request *pdpb.GetStoreRequest) (*pdpb.GetStoreResponse, error) {
+	if s.GetServiceMiddlewarePersistOptions().IsGRPCRateLimitEnabled() {
+		fName := currentFunction()
+		limiter := s.GetGRPCRateLimiter()
+		if limiter.Allow(fName) {
+			defer limiter.Release(fName)
+		} else {
+			return &pdpb.GetStoreResponse{
+				Header: s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN, errs.ErrReachRateLimit.FastGenByArgs().Error()),
```
Comment (nolouch) [inline server/grpc_service.go:778]: maybe we need to define an error type that can let the client to handle it.

File: server/grpc_service.go [hunk 3]
```diff
@@ -848,6 +871,17 @@ func (s *GrpcServer) PutStore(ctx context.Context, request *pdpb.PutStoreRequest
 
 // GetAllStores implements gRPC PDServer.
 func (s *GrpcServer) GetAllStores(ctx context.Context, request *pdpb.GetAllStoresRequest) (*pdpb.GetAllStoresResponse, error) {
```
Comment (CabinfeverB) [inline server/grpc_service.go:873]: How abou renaming `GetStores` to `GetAllStores` in http-api's router.go?

File: server/server.go [hunk 4]
```diff
@@ -299,8 +305,8 @@ func CreateServer(ctx context.Context, cfg *config.Config, services []string, le
 		diagnosticspb.RegisterDiagnosticsServer(gs, s)
 		// Register the micro services GRPC service.
 		s.registry.InstallAllGRPCServices(s, gs)
+		s.grpcServer = gs
```
Comment (CabinfeverB) [inline server/server.go:308]: When exec `initGRPCServiceLabels`, will it fetch the methods of ETCD gRPC server?

Comment (rleungx) [reply server/server.go:308]: Yes, we only have one gRPC server.

**Review by CabinfeverB — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Review by nolouch — APPROVED** [review]

**Pull request merged**